File: backend/routes/folders.py
from fastapi import APIRouter, HTTPException, Query

from filesystem import (
    FolderExplorerError,
    create_subfolder,
    get_home_folder,
    list_child_folders,
    list_folder_roots,
    open_folder_in_file_manager,
)
from folder_favorites import add_folder_favorite, list_folder_favorites, remove_folder_favorite
from routes._helpers import resolve_folder
from schemas import (
    FolderChildrenResponse,
    FolderCreateResponse,
    FolderFavoritesResponse,
    FolderOpenResponse,
    FolderRootsResponse,
)

router = APIRouter()


@router.get("/folders/roots", response_model=FolderRootsResponse)
def folder_roots() -> FolderRootsResponse:
    home = get_home_folder()
    return FolderRootsResponse(home=str(home), roots=list_folder_roots())


@router.get("/folders/children", response_model=FolderChildrenResponse)
def folder_children(
    path: str = Query(..., description="Absolute path to parent folder"),
) -> FolderChildrenResponse:
    """List immediate child folders only (no media items, stats, or last-folder side effects)."""
    folder = resolve_folder(path)
    return FolderChildrenResponse(folder=str(folder), children=list_child_folders(folder))


@router.get("/folders/favorites", response_model=FolderFavoritesResponse)
def read_folder_favorites() -> FolderFavoritesResponse:
    return FolderFavoritesResponse(favorites=list_folder_favorites())


@router.post("/folders/favorites", response_model=FolderFavoritesResponse)
def create_folder_favorite(
    path: str = Query(..., description="Absolute path to folder"),
) -> FolderFavoritesResponse:
    favorites = add_folder_favorite(path)
    return FolderFavoritesResponse(favorites=favorites)


@router.delete("/folders/favorites", response_model=FolderFavoritesResponse)
def delete_folder_favorite(
    path: str = Query(..., description="Absolute path to folder"),
) -> FolderFavoritesResponse:
    favorites = remove_folder_favorite(path)
    return FolderFavoritesResponse(favorites=favorites)


@router.post("/folders/open", response_model=FolderOpenResponse)
def open_folder_in_explorer(
    path: str = Query(..., description="Absolute path to folder"),
) -> FolderOpenResponse:
    folder = resolve_folder(path)

    try:
        open_folder_in_file_manager(folder)
    except FolderExplorerError as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc

    return FolderOpenResponse(path=str(folder))


@router.post("/folders/create", response_model=FolderCreateResponse)
def create_folder(
    path: str = Query(..., description="Absolute path to parent folder"),
    name: str = Query(..., description="Name for the new subfolder"),
) -> FolderCreateResponse:
    parent = resolve_folder(path)
    created = create_subfolder(parent, name)
    return FolderCreateResponse(**created)

File: frontend/src/features/browse/api/folders.ts
import { requestJson } from "@/shared/api/http";
import type {
  FolderChildrenResponse,
  FolderCreateResponse,
  FolderFavoritesResponse,
  FolderOpenResponse,
  FolderRootsResponse,
} from "@/shared/types";

export async function fetchFolderRoots(): Promise<FolderRootsResponse> {
  return requestJson<FolderRootsResponse>("/api/folders/roots");
}

/** Immediate child folders only — cheaper than full /api/browse for tree UIs. */
export async function fetchFolderChildren(folderPath: string): Promise<FolderChildrenResponse> {
  const params = new URLSearchParams({ path: folderPath });
  return requestJson<FolderChildrenResponse>(`/api/folders/children?${params}`);
}

export async function fetchFolderFavorites(): Promise<FolderFavoritesResponse> {
  return requestJson<FolderFavoritesResponse>("/api/folders/favorites");
}

export async function addFolderFavorite(folderPath: string): Promise<FolderFavoritesResponse> {
  const params = new URLSearchParams({ path: folderPath });
  return requestJson<FolderFavoritesResponse>(`/api/folders/favorites?${params}`, {
    method: "POST",
  });
}

export async function removeFolderFavorite(folderPath: string): Promise<FolderFavoritesResponse> {
  const params = new URLSearchParams({ path: folderPath });
  return requestJson<FolderFavoritesResponse>(`/api/folders/favorites?${params}`, {
    method: "DELETE",
  });
}

export async function openFolderInExplorer(folderPath: string): Promise<FolderOpenResponse> {
  const params = new URLSearchParams({ path: folderPath });
  return requestJson<FolderOpenResponse>(`/api/folders/open?${params}`, {
    method: "POST",
  });
}

export async function createFolder(
  parentPath: string,
  name: string,
): Promise<FolderCreateResponse> {
  const params = new URLSearchParams({ path: parentPath, name });
  return requestJson<FolderCreateResponse>(`/api/folders/create?${params}`, {
    method: "POST",
  });
}
